src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -333,725 +197,361 @@ bool G1BarrierSetC2::g1_can_remove_post_barrier(GraphKit* kit,
+  // It is OK to move the store across the object initialization boundary only
+  // if it does not have any barrier, or if it has barriers that can be safely
+  // elided (because of the compensation steps taken on the allocation slow path
+  // when ReduceInitialCardMarks is enabled).
+  return (MemNode::barrier_data(store) == 0) || use_ReduceInitialCardMarks();
 }
 
-#undef __
+void G1BarrierSetC2::clone_at_expansion(PhaseMacroExpand* phase, ArrayCopyNode* ac) const {
+  if (ac->is_clone_inst() && !use_ReduceInitialCardMarks()) {
+    clone_in_runtime(phase, ac, G1BarrierSetRuntime::clone_addr(), "G1BarrierSetRuntime::clone");
+    return;
+  }
+  BarrierSetC2::clone_at_expansion(phase, ac);
+}
 
-Node* G1BarrierSetC2::load_at_resolved(C2Access& access, const Type* val_type) const {
+Node* G1BarrierSetC2::store_at_resolved(C2Access& access, C2AccessValue& val) const {
   DecoratorSet decorators = access.decorators();
-  Node* adr = access.addr().node();
-  Node* obj = access.base();
-
-  bool anonymous = (decorators & C2_UNSAFE_ACCESS) != 0;
-  bool mismatched = (decorators & C2_MISMATCHED) != 0;
-  bool unknown = (decorators & ON_UNKNOWN_OOP_REF) != 0;
+  bool anonymous = (decorators & ON_UNKNOWN_OOP_REF) != 0;
   bool in_heap = (decorators & IN_HEAP) != 0;
-  bool in_native = (decorators & IN_NATIVE) != 0;
-  bool on_weak = (decorators & ON_WEAK_OOP_REF) != 0;
-  bool on_phantom = (decorators & ON_PHANTOM_OOP_REF) != 0;
-  bool is_unordered = (decorators & MO_UNORDERED) != 0;
-  bool no_keepalive = (decorators & AS_NO_KEEPALIVE) != 0;
-  bool is_mixed = !in_heap && !in_native;
-  bool need_cpu_mem_bar = !is_unordered || mismatched || is_mixed;
-
-  Node* top = Compile::current()->top();
-  Node* offset = adr->is_AddP() ? adr->in(AddPNode::Offset) : top;
-
-  // If we are reading the value of the referent field of a Reference
-  // object (either by using Unsafe directly or through reflection)
-  // then, if G1 is enabled, we need to record the referent in an
-  // SATB log buffer using the pre-barrier mechanism.
-  // Also we need to add memory barrier to prevent commoning reads
-  // from this field across safepoint since GC can change its value.
-  bool need_read_barrier = (((on_weak || on_phantom) && !no_keepalive) ||
-                            (in_heap && unknown && offset != top && obj != top));
+  bool tightly_coupled_alloc = (decorators & C2_TIGHTLY_COUPLED_ALLOC) != 0;
+  bool need_store_barrier = !(tightly_coupled_alloc && use_ReduceInitialCardMarks()) && (in_heap || anonymous);
+  if (access.is_oop() && need_store_barrier) {
+    access.set_barrier_data(get_store_barrier(access));
+    if (tightly_coupled_alloc) {
+      assert(!use_ReduceInitialCardMarks(),
+             "post-barriers are only needed for tightly-coupled initialization stores when ReduceInitialCardMarks is disabled");
+      access.set_barrier_data(access.barrier_data() ^ G1C2BarrierPre);
```

Comment:
I have been looking at this code after integration, and I wonder if `^` is really correct here? Was the intend to remove `G1C2BarrierPre` from the barrier data? What happens if `get_store_barrier` does not actually set it? Do we flip the bit back?

Reply:
Yes, the intend (and actual effect) is to remove `G1C2BarrierPre` from the barrier data. Using an XOR (`^`) is correct because at that program point `G1C2BarrierPre` is guaranteed to be set. This is because an `access` corresponding to a tightly-coupled initialization store is always of type `C2OptAccess`, hence `!access.is_parse_access()` and `get_store_barrier(access)` trivially returns `G1C2BarrierPre | G1C2BarrierPost`. Having said this, it would be clearly less convoluted to simply clear `G1C2BarrierPre` instead of flipping it. I will file a RFE, thanks.

As a side note, this complexity is necessary to handle `!ReduceInitialCardMarks`. I keep wondering if the benefit of being able to disable `ReduceInitialCardMarks` [1,2,3] is worth the significant complexity required in the GC-C2 interface to deal with it.

[1] https://docs.oracle.com/en/java/javase/23/gctuning/garbage-first-garbage-collector-tuning.html
[2] https://bugs.openjdk.org/browse/JDK-8166899
[3] https://bugs.openjdk.org/browse/JDK-8167077

Reply:
Reported here: [JDK-8341525](https://bugs.openjdk.org/browse/JDK-8341525).